### ISSUE TITLE
About the change and difference in the source code.

### ISSUE THREAD

**Issue by SunSwallow** [uchidalab/softmaxgradient-lrp#2]
Hello! I compared your source code and paper and find something that puzzled me.
The code we downloaded last year is a little different from that code now. 
In the "_head_mapping" function, the comments of  old code are:
 """
        target:yt，others:ytyj/(1-yt)
 """
But the comments of new code are:
"""
        target:yt(1-yt)  base: ytyi
"""
And according to the paper, I think it should be:
"""
        target:yt(1-yt)  base: -ytyi
"""
So I want to know why the code changed(because  I find that the old version is better for visualization), and whether there is something wrong in the current code.
Thanks!

**Comment by biwana** [comment]
target:yt，others:ytyj/(1-yt)

and 

target:yt(1-yt) base: ytyi

are the same thing, except the former has unnecessary normalization (multiply both sides by 1-yt )

As for -ytyi, in class _LRPSubtraction() , the "target" is subtracted from the "base".  so the code says positive ytyi, but later it becomes -ytyi

This is different from the sequential version because the sequential version doesn't rely on _LRPSubtraction()

**Comment by SunSwallow** [comment]
Thanks very much for your answering! Your answer is very useful. I will do some more experiments. Thanks again~

**Comment by SunSwallow** [comment]
Sorry...There is still one little question.
In the "_head_mapping" function, the following two writings are equivalent:

"""  target:yt(1-yt) base: ytyi  """   
X = Lambda(
            lambda x: (x * (1.- target_value[:, None] + epsilon)) * Kronecker_delta + x * Inv_Kronecker_delta * target_value[:, None],
            output_shape=lambda input_shape: (None, int(input_shape[1])))(X)

and

"""  target:yt，others:ytyj/(1-yt) """
X = Lambda(
            lambda x: x * Kronecker_delta + x * Inv_Kronecker_delta * target_value[:, None] / (1.- target_value[:, None] + epsilon),
            output_shape=lambda input_shape: (None, int(input_shape[1])))(X)

When the target_value is big (close to 1), the heatmaps generated by these two methods look like the same, while when the target_value is small (close to 0, which means that the target value is not the maximum value in the softmax vector), two heatmaps are different. I want to know how the normalization part affects the generation of heatmap. I think when target_value is small, the normalization part  (1.- target_value[:, None] + epsilon) is close to 1 and shouldn't cause the change of heatmaps. It's really puzzled.
Thanks!

**Comment by biwana** [comment]
Technically, they will have different values. But, I found that it doesn't matter because they are different by a prediction-wise constant (1-yt). The reason it doesn't matter is because it is standard practice to normalize the heatmap anyway. For example, in my helper.py the output maps are normalized before visualization. So, there is no difference between the two after the normalization in the heatmap visualization. The same for if you are calculating something like AOPC, because all of the result values are scaled by a constant. 

To be honest, the original function was programmed by my co-author, and I don't know why he chose the old method of calculating.  I changed it to the current method for simplicity and because it doesn't have a problem with yt = 1.
